amaru/src/consensus/header_validation.rs

```diff
@@ -0,0 +1,179 @@
+use crate::{
+    consensus::{Point, ValidateHeaderEvent},
+    sync::PullEvent,
+};
+use gasket::framework::*;
+use miette::miette;
+use ouroboros::{ledger::LedgerState, validator::Validator};
+use ouroboros_praos::consensus::BlockValidator;
+use pallas_crypto::hash::Hash;
+use pallas_math::math::{FixedDecimal, FixedPrecision};
+use pallas_network::facades::PeerClient;
+use pallas_primitives::conway::Epoch;
+use pallas_traverse::MultiEraHeader;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use tracing::{info, trace};
+
+pub type UpstreamPort = gasket::messaging::InputPort<PullEvent>;
+pub type DownstreamPort = gasket::messaging::OutputPort<ValidateHeaderEvent>;
+
+/// Mocking this calculation specifically for preprod. We need to know the epoch for the slot so
+/// we can look up the epoch nonce.
+fn epoch_for_preprod_slot(slot: u64) -> u64 {
+    let shelley_epoch_length = 432000; // 5 days in seconds
+    let shelley_transition_epoch: u64 = 4;
+    let byron_protocol_consts_k: u64 = 2160;
+    let byron_epoch_length = 10 * byron_protocol_consts_k;
+    let byron_slots = byron_epoch_length * shelley_transition_epoch;
+    let shelley_slots = slot - byron_slots;
+    let epoch = (shelley_slots / shelley_epoch_length) + shelley_transition_epoch;
+
+    epoch
+}
```

Comment:
@Quantumplation here's something that can be its own stream of work: slotting arithmetic. In the long run, we will need code that can perform a variety of conversion from slots, epochs, posix times, etc.. from (evolving) protocol parameters and multi-era sizes. In particular, it should also handle notions of forecast into the future given a certain stability window's size. 

I believe this to be something sufficiently well-scoped (I can spend more time drafting that) and isolated that it can even live in its own crate. Is something Sundae would be interested in assisting with?

Reply:
Yes, absolutely!

Reply:
This work is already done inside of cncli. It probably needs to be cleaned up, but the arithmetic is all there.

Reply:
https://github.com/cardano-community/cncli/blob/develop/src/nodeclient/leaderlog/mod.rs#L200

https://bitbucket.org/muamw10/jormanager/src/caa5b601c70b4836219000c3990bd1979e87f11e/src/main/kotlin/com/swiftmako/jormanager/controllers/utils/BlockUtils.kt#lines-58

Reply:
@AndrewWestberg, I was hinting at something quite more sophisticated here. Since the problem we have isn't really to perform slot arithmetic from known parameters (that's indeed straightforward), but to provide an abstraction that (1) keeps track of the ever evolving parameters, (2) ensures that we follow heuristic similar to the Haskell implementation (e.g. no time conversation in the future outside of the stability window).